### ISSUE TITLE
truncate! ruins QN conserving MPS

### ISSUE THREAD

**Issue by jtschneider** [ITensor/ITensors.jl#619]
Hello guys,

**tldr;** When trying to truncate a QN-conserving MPS below the minimal required linkdim to conserve it's total charge, we encounter an error that 'ruins' the state.

I encountered a strange bug that I am not sure of whether it is intended. When working with a QN conserving MPS I can go on and about to truncate its linkdim via `truncate!(psi; maxdim=m)` where `m` is anything less than `maxlinkdim(psi)`.
Let me consider the ground state on the Heisenberg chain with spin-1/2,
```julia 
N=12
ampo = AutoMPO()
    for j in 1:N-1
        ampo .+= 0.5, "S+", j, "S-", j+1
        ampo .+= 0.5, "S-", j, "S+", j+1
        ampo .+= 1.0, "Sz", j, "Sz", j+1
    end

sites = siteinds("S=1/2",N,conserve_qns=true)

H = MPO(ampo, sites)
state = [isodd(n) ? "Up" : "Dn" for n in 1:N]
psirand = randomMPS(sites, state, 6);
sweeps = Sweeps(8)
maxdim!(sweeps, 10, 20, 40)
energy0, psi0 = dmrg(H, psirand, sweeps; outputlevel=0)
```
I now want to truncate this ground state to approximate the 'closest product state'. Truncating to `linkdim=1` should not work
```julia
julia> ψ = deepcopy(psi0)

julia> truncate!(ψ; maxdim=1)
ERROR: BoundsError: attempt to access 0-element Vector{Pair{QN, Int64}} at index [1]
Stacktrace:
  [1] getindex
    @ ./array.jl:801 [inlined]
  [2] combineblocks(qns::Vector{Pair{QN, Int64}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/qn/qnindex.jl:302
  [3] combineblocks
    @ ~/.julia/packages/ITensors/Ligbx/src/qn/qnindex.jl:356 [inlined]
  [4] combiner(inds::IndexSet{1, Index{Vector{Pair{QN, Int64}}}, Tuple{Index{Vector{Pair{QN, Int64}}}}}; kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/qn/qnitensor.jl:166
  [5] combiner
    @ ~/.julia/packages/ITensors/Ligbx/src/qn/qnitensor.jl:163 [inlined]
  [6] #combiner#177
    @ ~/.julia/packages/ITensors/Ligbx/src/itensor.jl:1219 [inlined]
  [7] combiner(inds::Index{Vector{Pair{QN, Int64}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/itensor.jl:1219
  [8] svd(A::ITensor{3}, Linds::IndexSet{2, Index{Vector{Pair{QN, Int64}}}, Tuple{Index{Vector{Pair{QN, Int64}}}, Index{Vector{Pair{QN, Int64}}}}}; kwargs::Base.Iterators.Pairs{Symbol, Any, Tuple{Symbol, Symbol}, NamedTuple{(:tags, :alg), Tuple{TagSet, String}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/decomp.jl:72
  [9] factorize_svd(A::ITensor{3}, Linds::IndexSet{2, Index{Vector{Pair{QN, Int64}}}, Tuple{Index{Vector{Pair{QN, Int64}}}, Index{Vector{Pair{QN, Int64}}}}}; kwargs::Base.Iterators.Pairs{Symbol, TagSet, Tuple{Symbol}, NamedTuple{(:tags,), Tuple{TagSet}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/decomp.jl:298
 [10] factorize(A::ITensor{3}, Linds::IndexSet{2, Index{Vector{Pair{QN, Int64}}}, Tuple{Index{Vector{Pair{QN, Int64}}}, Index{Vector{Pair{QN, Int64}}}}}; kwargs::Base.Iterators.Pairs{Symbol, TagSet, Tuple{Symbol}, NamedTuple{(:tags,), Tuple{TagSet}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/decomp.jl:414
 [11] orthogonalize!(M::MPS, j::Int64; kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/mps/abstractmps.jl:1086
 [12] orthogonalize!
    @ ~/.julia/packages/ITensors/Ligbx/src/mps/abstractmps.jl:1076 [inlined]
 [13] truncate!(M::MPS; kwargs::Base.Iterators.Pairs{Symbol, Int64, Tuple{Symbol}, NamedTuple{(:maxdim,), Tuple{Int64}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/mps/abstractmps.jl:1138
 [14] top-level scope
    @ REPL[135]:1
```
as the QN-conserving MPS needs to have a constant total spin of 0 as defined by the conserved charge sector of `state =  [isodd(n) ? "Up" : "Dn" for n in 1:N]` (`N` being even) and the flow across the MPS with linkdim=1 would not match that. However, `maxdim=2` should work, but it does not when I tried `maxdim=1` before!
```julia 
julia> truncate!(ψ; maxdim=2)
ERROR: BoundsError: attempt to access 0-element Vector{Pair{QN, Int64}} at index [1]
Stacktrace:
  [1] getindex
    @ ./array.jl:801 [inlined]
  [2] combineblocks(qns::Vector{Pair{QN, Int64}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/qn/qnindex.jl:302
  [3] combineblocks
    @ ~/.julia/packages/ITensors/Ligbx/src/qn/qnindex.jl:356 [inlined]
  [4] combiner(inds::IndexSet{1, Index{Vector{Pair{QN, Int64}}}, Tuple{Index{Vector{Pair{QN, Int64}}}}}; kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/qn/qnitensor.jl:166
  [5] combiner
    @ ~/.julia/packages/ITensors/Ligbx/src/qn/qnitensor.jl:163 [inlined]
  [6] #combiner#177
    @ ~/.julia/packages/ITensors/Ligbx/src/itensor.jl:1219 [inlined]
  [7] combiner(inds::Index{Vector{Pair{QN, Int64}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/itensor.jl:1219
  [8] svd(A::ITensor{3}, Linds::IndexSet{2, Index{Vector{Pair{QN, Int64}}}, Tuple{Index{Vector{Pair{QN, Int64}}}, Index{Vector{Pair{QN, Int64}}}}}; kwargs::Base.Iterators.Pairs{Symbol, Any, Tuple{Symbol, Symbol}, NamedTuple{(:tags, :alg), Tuple{TagSet, String}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/decomp.jl:72
  [9] factorize_svd(A::ITensor{3}, Linds::IndexSet{2, Index{Vector{Pair{QN, Int64}}}, Tuple{Index{Vector{Pair{QN, Int64}}}, Index{Vector{Pair{QN, Int64}}}}}; kwargs::Base.Iterators.Pairs{Symbol, TagSet, Tuple{Symbol}, NamedTuple{(:tags,), Tuple{TagSet}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/decomp.jl:298
 [10] factorize(A::ITensor{3}, Linds::IndexSet{2, Index{Vector{Pair{QN, Int64}}}, Tuple{Index{Vector{Pair{QN, Int64}}}, Index{Vector{Pair{QN, Int64}}}}}; kwargs::Base.Iterators.Pairs{Symbol, TagSet, Tuple{Symbol}, NamedTuple{(:tags,), Tuple{TagSet}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/decomp.jl:414
 [11] orthogonalize!(M::MPS, j::Int64; kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/mps/abstractmps.jl:1086
 [12] orthogonalize!
    @ ~/.julia/packages/ITensors/Ligbx/src/mps/abstractmps.jl:1076 [inlined]
 [13] truncate!(M::MPS; kwargs::Base.Iterators.Pairs{Symbol, Int64, Tuple{Symbol}, NamedTuple{(:maxdim,), Tuple{Int64}}})
    @ ITensors ~/.julia/packages/ITensors/Ligbx/src/mps/abstractmps.jl:1138
 [14] top-level scope
    @ REPL[135]:1
```
It only works when I deepcopy the original state again:
```julia
julia> ψ = deepcopy(psi0)

julia> m = 2

julia> truncate!(ψ;maxdim=m)
MPS
[1] IndexSet{2} (dim=2|id=527|"S=1/2,Site,n=1") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=2|id=629|"Link,l=1") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1

[2] IndexSet{3} (dim=2|id=421|"S=1/2,Site,n=2") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=1|id=18|"Link,l=2") <Out>
 1: QN("Sz",0) => 1
(dim=2|id=629|"Link,l=1") <In>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1

[3] IndexSet{3} (dim=2|id=168|"S=1/2,Site,n=3") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=2|id=803|"Link,l=3") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=1|id=18|"Link,l=2") <In>
 1: QN("Sz",0) => 1

[4] IndexSet{3} (dim=2|id=357|"S=1/2,Site,n=4") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=1|id=66|"Link,l=4") <Out>
 1: QN("Sz",0) => 1
(dim=2|id=803|"Link,l=3") <In>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1

[5] IndexSet{3} (dim=2|id=35|"S=1/2,Site,n=5") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=2|id=127|"Link,l=5") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=1|id=66|"Link,l=4") <In>
 1: QN("Sz",0) => 1

[6] IndexSet{3} (dim=2|id=854|"S=1/2,Site,n=6") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=1|id=881|"Link,l=6") <Out>
 1: QN("Sz",0) => 1
(dim=2|id=127|"Link,l=5") <In>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1

[7] IndexSet{3} (dim=2|id=277|"S=1/2,Site,n=7") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=2|id=649|"Link,l=7") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=1|id=881|"Link,l=6") <In>
 1: QN("Sz",0) => 1

[8] IndexSet{3} (dim=2|id=682|"S=1/2,Site,n=8") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=1|id=857|"Link,l=8") <Out>
 1: QN("Sz",0) => 1
(dim=2|id=649|"Link,l=7") <In>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1

[9] IndexSet{3} (dim=2|id=190|"S=1/2,Site,n=9") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=2|id=541|"Link,l=9") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=1|id=857|"Link,l=8") <In>
 1: QN("Sz",0) => 1

[10] IndexSet{3} (dim=2|id=536|"S=1/2,Site,n=10") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=1|id=902|"Link,l=10") <Out>
 1: QN("Sz",0) => 1
(dim=2|id=541|"Link,l=9") <In>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1

[11] IndexSet{3} (dim=2|id=94|"S=1/2,Site,n=11") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=2|id=537|"Link,l=11") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=1|id=902|"Link,l=10") <In>
 1: QN("Sz",0) => 1

[12] IndexSet{2} (dim=2|id=278|"S=1/2,Site,n=12") <Out>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
(dim=2|id=537|"Link,l=11") <In>
 1: QN("Sz",1) => 1
 2: QN("Sz",-1) => 1
 ```
Note this works with any `m>=2`; including the occurence of the bug.

**Comment by jtschneider** [comment]
:facepalm: well an easy workaround is the function `truncate(ψ; maxdim=m)` (not-inplace) :sweat_smile:

**Comment by mtfishman** [comment]
Right, I'm not sure we can -- or should -- do much about this kind of thing, it would really only be an issue when using it interactively. If you weren't using it interactively/in a global scope, your run would fail and there wouldn't be an opportunity to try to use that MPS again.

For example this is a general "problem" for any in-place Julia function:
```julia
julia> function f!(x)
         for n in eachindex(x)
           n > length(x) ÷ 2 && error("Stop")
           x[n] = 0
         end
       end
f! (generic function with 1 method)

julia> v = randn(10);

julia> f!(v)
ERROR: Stop
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] f!(x::Vector{Float64})
   @ Main ./REPL[57]:3
 [3] top-level scope
   @ REPL[59]:1

julia> v
10-element Vector{Float64}:
  0.0
  0.0
  0.0
  0.0
  0.0
 -1.2731340649172502
 -0.18886975247025578
 -1.5946096078551426
 -0.11676854782278846
  1.5127250495824236
```
To "fix" this in general you would need to store the original storage, catch if there was an error, and then return the original storage, which would ruin the point of using in-place operations.

In general, I prefer to use the out-of-place versions of MPS functions since it makes the code logic simpler, and there is often no efficiency gain to using the in-place MPS versions. The out-of-place MPS versions don't do "deep copies" of the MPS, so they don't allocate any more ITensor data than is needed. In general they just copy the vector that points to the same ITensor data and only allocates ITensor data as needed for the operation. Note there may be some exceptions to this, like if we introduce a function `normalize!(::MPS)` that scaled individual tensors of the MPS, but in general there is practically no efficiency gain for using in-place MPS functions.
